lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -244,6 +244,90 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
       decls.push_back(cast<TypeDecl>(found.getValueDecl()));
   }
 
+  // If we're looking for a type named `CodingKeys`, then we've got some
+  // additional work to do here. CodingKeys is a special type: although not a
+  // protocol requirement of the Encodable/Decodable protocols, it is
+  // synthesized as part of Encodable/Decodable conformance. Unqualified lookup
+  // may occur before the type is synthesized, though, in which case, we need to
+  // make sure it gets synthesized.
+  //
+  // The unqualified lookups above, however, cannot force Encodable/Decodable
+  // synthesis to occur (since they know nothing about typechecking). Thus, we
+  // also need to find the closest qualified type which might contain a
+  // CodingKeys type and synthesize it.
+  //
+  // If this type is "closer" than the closest unqualified type, then we can
+  // use it.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
+    SmallVector<TypeDecl *, 1> qualifiedDecls;
+
+    auto *currentDC = dc;
+    while (currentDC) {
+      auto *nominal =
+        currentDC->getAsNominalTypeOrNominalTypeExtensionContext();
+      if (nominal) {
+        // We've found a nominal context to look in. Let's preform a qualified
+        // lookup here (which will cause CodingKeys synthesis if possible).
+        auto lookupOptions = options | NameLookupFlags::ProtocolMembers;
+        auto results = lookupMemberType(nominal,
+                                        nominal->getDeclaredInterfaceType(),
+                                        Context.Id_CodingKeys, lookupOptions);
+        if (results.size() > 0) {
+          for (auto &result : results) {
+            qualifiedDecls.push_back(result.first);
+          }
+
+          break;
+        }
+      }
+
+      currentDC = currentDC->getParent();
+    }
+
+    // If there were any results, we need to merge them with any decls already
+    // found by the unqualified lookup above. Those decls might take precedence
+    // based on their locality.
```

Comment:
Yeah... Again, putting this logic into `DeclContext::lookupQualified` would be ideal since both qualified and unqualified lookups go through that path, but there's no knowledge in there about `TypeChecker`.

About the "level" — is it safe, then, to look at the first element in `decls` and `qualifiedDecls` and return the vector containing whichever is closer?